### PR TITLE
[KBFS-4014] offline sync settings UI

### PR DESCRIPTION
**Pull request by jzila**
Tagging kbfs-hackers since there are some KBFS fixes as well that need review. Ideally @songgao and/or @jakob223 would take a look.

This PR has draft code for the mobile settings page, but that's not complete. I'll open a new PR for that.

### CODE REVIEW

File: shared/common-adapters/dropdown.js [hunk 1]
```diff
@@ -84,7 +86,9 @@ class Dropdown<N: React.Node> extends React.Component<Props<N> & OverlayParentPr
             {this.props.items.map((i, idx) => (
               <ClickableBox
                 key={idx}
-                onClick={() => {
+                onClick={(evt: SyntheticEvent<>) => {
+                  evt.stopPropagation && evt.stopPropagation()
+                  evt.preventDefault && evt.preventDefault()
```
Comment (jzila) [inline shared/common-adapters/dropdown.js]: @buoyad I'm having trouble getting this to actually prevent default. Any suggestions?

File: shared/settings/files/index.js [hunk 2]
```diff
@@ -1,28 +1,110 @@
...
+  spaceAvailableNotificationThreshold: number,
+  onChangedSyncNotifications: (number) => void,
+  onEnableSyncNotifications: () => void,
+  onDisableSyncNotifications: () => void,
 |}
 
+export const allowedNotificationThresholds = [
+  100 * 1024 ** 2,
+  1024 ** 3,
+  10 * 1024 ** 3,
```
Comment (jakob223) [inline shared/settings/files/index.js]: maybe add one more option in like the 2-5 GB range? I feel like I'd want to set it somewhere around there personally.

File: shared/settings/files/index.js [hunk 3]
```diff
@@ -1,28 +1,110 @@
...
 import RefreshDriverStatusOnMount from '../../fs/common/refresh-driver-status-on-mount'
+import RefreshSettings from './refresh-settings'
+import flags from '../../util/feature-flags'
 
 type Props = {|
+  areSettingsLoading: boolean,
   driverStatus: Types.DriverStatus,
-  onEnable: () => void,
   onDisable: () => void,
+  onEnable: () => void,
```
Comment (jakob223) [inline shared/settings/files/index.js]: I'm confused why you moved this down - if it's for alphabetical order you should probably swap `onEnableSyncNotifications` down too.

File: shared/settings/files/index.js [hunk 4]
```diff
@@ -1,28 +1,110 @@
...
+    selectedBoxStyle={styles.syncNotificationDropdownItem}
+    disabled={props.areSettingsLoading || props.spaceAvailableNotificationThreshold === 0}
+  />
+)
+
+const ThresholdDropdown = Platform.isMobile ? ThresholdDropdownMobile : ThresholdDropdownDesktop
+
+const SyncNotificationSetting = (props: Props) => (
+  <Kb.Box2 direction="horizontal" alignItems="center">
+    <Kb.Text type="Body">Warn me if I only have less than </Kb.Text>
```
Comment (jakob223) [inline shared/settings/files/index.js]: "only have less than" feels redundant to me. Maybe cut the 'only'?

File: shared/settings/files/index.js [hunk 5]
```diff
@@ -1,28 +1,110 @@
...
+          onDone={this._hide}
+          onSelect={this._select}
+        />
+      </>
+    )
+  }
+}
+
+const ThresholdDropdownDesktop = (props: Props) => (
+  <Kb.Dropdown
```
Comment (jakob223) [inline shared/settings/files/index.js]: The place the dropdown appears (not actually below the dropdown box, but centered on it) feels weird to me.

Also, the text in it is way smaller than the box itself - it feels to me like the expanded box should be the same size as the box pre-expansion. If it stays big, the text should probably be centered in it.

Finally, there should be margins around the text in the dropped-down box - right now the text is touching the borders.

no margins

File: shared/settings/files/index.js [hunk 6]
```diff
@@ -71,12 +175,30 @@ const styles = Styles.styleSheetCreate({
...
   },
+  syncContent: {
+    paddingLeft: Styles.globalMargins.xsmall,
+    paddingTop: Styles.globalMargins.medium,
+  },
+  syncNotificationCheckbox: {
+    alignItems: 'center',
+  },
+  syncNotificationDropdownItem: {
+    width: 2 * Styles.globalMargins.xlarge,
```
Comment (jakob223) [inline shared/settings/files/index.js]: why is this so big?

File: shared/settings/files/index.js (hunk 4 above)
Comment (jzila) [reply shared/settings/files/index.js]: Copy is from design. I'll change it though to "Warn me if I have less than <n> storage space remaining".

File: shared/settings/files/index.js (hunk 6 above)
Comment (jzila) [reply shared/settings/files/index.js]: It's actually pretty small, that's how we get the root dropdown to look so small.

File: shared/settings/files/index.js (hunk 5 above)
Comment (jzila) [reply shared/settings/files/index.js]: I was having trouble figuring out how to style this.

**Comment by jzila** [comment]
@jakob223 fixes implemented.

File: shared/settings/files/index.tsx [hunk 7]
```diff
@@ -202,7 +203,11 @@ const styles = Styles.styleSheetCreate({
     alignItems: 'center',
   },
   syncNotificationDropdownItem: {
-    width: 2 * Styles.globalMargins.xlarge,
+    alignItems: 'center',
+    paddingLeft: Styles.globalMargins.small,
+  },
+  syncNotificationDropdownOverlay: {
+    width: Styles.globalMargins.xlarge + Styles.globalMargins.medium,
```
Comment (jakob223) [inline shared/settings/files/index.tsx]: Why is this expressed as a sum of margins instead of as a number?

Comment (jzila) [reply shared/settings/files/index.tsx]: I always try to express dimensions as linear combinations of margins so it looks "good" relative to all the other dimensions we use, but usually I change it to an actual amount after the fact. I can do that here.

**Pull request merged**